### PR TITLE
Newline before macro var fix

### LINKED ISSUE #1031: Multi-line macro object definition triggers error recovery
```
DEFINE_FLOORS(carpet/regalcarpet,
	name = "regal carpet";\
	icon = 'icons/turf/floors.dmi';\
	icon_state = "regal_carpet";\
	step_material = "step_carpet";\
	step_priority = STEP_PRIORITY_MED)
``` 
with
```
#define DEFINE_FLOORS(_PATH, _VARS) \
	/turf/simulated/floor/_PATH{_VARS};\
	/turf/unsimulated/floor/_PATH{_VARS};\
	/turf/simulated/floor/airless/_PATH{_VARS};\
	/turf/unsimulated/floor/airless/_PATH{_VARS};
```

causes error recovery, instead of proper object definitions. Possibly because multiple objects are defined on one line?

### PR DESCRIPTION
**Pull request by amylizzle**
Fixes #1031

### CODE REVIEW

File: Content.Tests/DMProject/Tests/Preprocessor/macro_newline_var.dm [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#define GLOBAL_LIST_INIT(X, InitValue) GLOBAL_RAW(/list/##X); GLOBAL_MANAGED(X, InitValue)
+
+DEFINE_FLOORS(carpet/regalcarpet,
+	name = "regal carpet";\
+	icon = 'icons/obj/floors.dmi';\
+	icon_state = "regal_carpet";)
+
+DEFINE_FLOORS(carpet/regalcarpet/border,
+	name = "regal carpet border";\
+	icon = 'icons/obj/floors.dmi';\
```
Comment (ZeWaka) [inline Content.Tests/DMProject/Tests/Preprocessor/macro_newline_var.dm:24]: i mean, we don't have that icon file right? probably should change the tests slightly

Comment (amylizzle) [reply Content.Tests/DMProject/Tests/Preprocessor/macro_newline_var.dm:24]: fair point, although shouldn't we be erroring on that?

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessor.cs [hunk 2]
```diff
@@ -747,7 +747,12 @@ private List<Token> GetLineOfTokens()
 
             int parenthesisNesting = 1;
             while(true) {
+
                 switch (parameterToken.Type) {
+                    case TokenType.Newline:
+                        currentParameter.Add(new Token(TokenType.DM_Preproc_LineSplice, "", parameterToken.Location, null));
+                        parameterToken = GetNextToken(true);
+                        continue;
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessor.cs:755]: This does not compile in BYOND:
```
#define TEST(a,b) a+b

/proc/main()
    world.log << TEST(1, 2+
2)
```

This implies that you shouldn't just remove every newline you encounter. I think what BYOND is doing here is consuming an optional newline and/or whitespace after the comma separating the arguments.

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessor.cs [hunk 3]
```diff
@@ -754,6 +754,13 @@ private List<Token> GetLineOfTokens()
                         parameters.Add(currentParameter);
                         currentParameter = new List<Token>();
                         parameterToken = GetNextToken(true);
+                        while(parameterToken.Type == TokenType.Newline || parameterToken.Type == TokenType.DM_Whitespace) {
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessor.cs:757]: parameterToken is already gotten with `GetNextToken(true)` which should ignore whitespace tokens, making the check for `TokenType.DM_Whitespace` here unneeded.

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessor.cs [hunk 4]
```diff
@@ -754,6 +754,13 @@ private List<Token> GetLineOfTokens()
...
+                            if(parameterToken.Type == TokenType.Newline)
+                                currentParameter.Add(new Token(TokenType.DM_Preproc_LineSplice, "", parameterToken.Location, null));
+                            else
+                                currentParameter.Add(parameterToken);
+                            parameterToken = GetNextToken();
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessor.cs:762]:
```suggestion
                            parameterToken = GetNextToken(true);
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**